File: lintcode/python/0808_movie_network.py
#!/usr/bin/python -t

# bfs and heap
# AC


import heapq, collections

class Solution:
    """
    @param rating: the rating of the movies
    @param G: the realtionship of movies
    @param S: the begin movie
    @param K: top K rating 
    @return: the top k largest rating moive which contact with S
    """
    def topKMovie(self, rating, G, S, K):
        # Write your code here
        related_movies = self.bfs(G, S)
        #print related_movies
        
        heap = []
        ret = []
        
        for m in related_movies:
            if len(heap) >= K:
                heapq.heappushpop(heap, (rating[m], m))
            else:
                heapq.heappush(heap, (rating[m], m))
        
        #print heap
        
        ret = sorted([m for _, m in heap])
        
        return ret
        
        
    def bfs(self, g, s):
        q = collections.deque([s])
        v = set([s])
        ret = []
        
        while len(q) != 0:
            movie = q.popleft()
            ret.append(movie)
            for m in g[movie]:
                if m in v:
                    continue
                q.append(m)
                v.add(m)
                
        return ret[1:]
            

# BFS with heap


import heapq
from collections import deque

class Solution:
    """
    @param rating: the rating of the movies
    @param G: the realtionship of movies
    @param S: the begin movie
    @param K: top K rating 
    @return: the top k largest rating moive which contact with S
    """
    def topKMovie(self, rating, G, S, K):
        # Write your code here
        q = deque()
        h = []
        v = set()
        ret = []
        
        q.append(S)
        v.add(S)
        
        while len(q) > 0:
            cur = q.popleft()
            
            for r in G[cur]:
                if r in v:
                    continue
                if len(h) < K:
                    heapq.heappush(h, [rating[r], r])
                else:
                    if rating[r] > h[0][0]:
                        heapq.heappop(h)
                        heapq.heappush(h, [rating[r], r])
                       
                v.add(r)
                q.append(r)
        
        #print h        
        for ele in h:
            ret.append(ele[1])
         
        #print ret   
        return ret
        

File: lintcode/golang/0808_movie_network.go

// heap + bfs + sort

import (
	"container/heap"
	"sort"
)

type Node struct {
	rate int
	num  int
}

type IntHeap []*Node

func (h IntHeap) Len() int           { return len(h) }
func (h IntHeap) Less(i, j int) bool { return h[i].rate < h[j].rate }
func (h IntHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *IntHeap) Push(x interface{}) {
	// Push and Pop use pointer receivers because they modify the slice's length,
	// not just its contents.
	*h = append(*h, x.(*Node))
}

func (h *IntHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

/**
 * @param rating: the rating of the movies
 * @param G: the realtionship of movies
 * @param S: the begin movie
 * @param K: top K rating
 * @return: the top k largest rating moive which contact with S
 */
func topKMovie(rating []int, G [][]int, S int, K int) []int {
	// Write your code here
	relatedM := getRelated(G, S)

	minHeap := &IntHeap{}
	ret := []int{}

	for _, m := range relatedM {
		heap.Push(minHeap, &Node{rate: rating[m], num: m})
		if minHeap.Len() > K {
			heap.Pop(minHeap)
		}
	}

	sort.Slice(*minHeap, func(i, j int) bool {
		return []*Node(*minHeap)[i].rate < []*Node(*minHeap)[j].rate
	})

	for _, node := range []*Node(*minHeap) {
		ret = append(ret, node.num)
	}

    sort.Ints(ret)
	return ret
}

func getRelated(g [][]int, s int) []int {
	v := make(map[int]bool)
	q := []int{}
	q = append(q, s)
	v[s] = true
	ret := []int{}

	for len(q) > 0 {
		m := q[0]
		q = q[1:]
		if m != s {
			ret = append(ret, m)
		}
		for _, nei := range g[m] {
			if v[nei] != true {
				q = append(q, nei)
				v[nei] = true
			}
		}
	}

	return ret
}
